// capsolver-npm/src/CapSolver.js
const Tasker = require('./Tasker')
const axios = require('axios')

/**
 * CapSolver.com Tasks Handler
 */
class CapSolver {
    constructor(apikey, verbose=0, rqdelay=1700) { this.apikey = apikey; this.verbose = verbose; this.rqdelay = rqdelay; this.init() }

    /** * Set-up handler **/
    init(){ if(this.verbose === 2){ console.log(`capsolve-npm: Verbose level ${this.verbose} running at: ${this.apikey}`) } }

    /** * Return USD balance as float number **/
    async balance(){ let handled = await this.getBalance(); return handled['apiResponse']['balance'] ? parseFloat(handled['apiResponse']['balance']): handled }

    /**
     * One request to api.capsolver.com/getBalance
     */
    async getBalance(){
        let self = this
        let handled = await axios({ method: 'post', url: 'https://api.capsolver.com/getBalance', headers: { }, data: { 'clientKey': this.apikey } })
            .then(function (response) {
                if(self.verbose === 2){ console.log(response.data) }
                if(response.data['errorId'] !== 0){ return { 'error':-1, 'statusText':response.status, 'apiResponse':response.data } }
                return { 'error':0, 'statusText':response.status, 'apiResponse':response.data }
            })
            .catch(function (error) {
                if(error.response === undefined){ return error }
                if(self.verbose === 2){ console.log(error) }
                return { 'error':-1, 'statusText':error.response.status, 'apiResponse':error.response.data }
            })
        if(this.verbose !== 0) { console.log(`capsolver-npm: Balance ${parseFloat(handled['apiResponse']['balance'])} USD]`) }
        return handled
    }

    /**
     * Appends specific proxy connection details to taskData
     *
     * @param {object} tasker - tasker instance
     * @param {object} proxyInfo - proxy connection details schema
     */
    attachProxy(tasker, proxyInfo){
        if(proxyInfo.hasOwnProperty('proxy')){ tasker.taskData.proxy = proxyInfo.proxy }else{
            if(proxyInfo.proxyType !== null || true){ tasker.taskData.proxyType = proxyInfo.proxyType }
            tasker.taskData.proxyAddress = proxyInfo.proxyAddress
            tasker.taskData.proxyPort = proxyInfo.proxyPort
            if(proxyInfo.proxyLogin !== null || true){ tasker.taskData.proxyLogin = proxyInfo.proxyLogin }
            if(proxyInfo.proxyPassword !== null || true){ tasker.taskData.proxyPassword = proxyInfo.proxyPassword }
        }
        return tasker
    }

    /**
     * Handle results for one specific captcha task
     *
     * @param {object} taskData - taskData schema
     * @param retrieve
     */
    async runAnyTask(taskData, retrieve=true) {
        if(taskData.hasOwnProperty('type')){
            let tasker = new Tasker(null, this.apikey, this.verbose, retrieve)
            tasker.taskData = taskData
            if(taskData.hasOwnProperty('proxyInfo')){ this.attachProxy(tasker, taskData.proxyInfo) }
            let tasked = await tasker.createTask()
            if(tasked.error !== 0) return tasked
            return await tasker.getTaskResult(tasked['apiResponse']['taskId'], this.rqdelay)
        }else{
            throw TypeError('taskData has not type property.')
        }
    }

    /** task-binded methods **/

    /** mtcaptcha **/
    async mtcaptcha(websiteURL, websiteKey, proxyInfo){
        let tasker = new Tasker('MtCaptchaTask', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    /** img2txt **/
    async mtcaptchaproxyless(websiteURL, websiteKey){
        let tasker = new Tasker('MtCaptchaTaskProxyLess', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        return await tasker.execute(this.rqdelay)
    }

    /** img2txt **/
    async image2text(body){
        let tasker = new Tasker('ImageToTextTask', this.apikey, this.verbose, false)
        // binding
        tasker.taskData.body = body
        return await tasker.execute(this.rqdelay)
    }

    /** hcap **/
    async hcaptcha(websiteURL, websiteKey, proxyInfo, userAgent=null, isInvisible=null, enterprisePayload=null){
        let tasker = new Tasker('HCaptchaTask', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(isInvisible!==null) { tasker.taskData.isInvisible = true }
        if(enterprisePayload!==null) { tasker.taskData.isEnterprise = true; tasker.taskData.enterprisePayload = enterprisePayload }
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    async hcaptchaproxyless(websiteURL, websiteKey, userAgent=null, isInvisible=null, enterprisePayload=null){
        let tasker = new Tasker('HCaptchaTaskProxyLess', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(isInvisible!==null) { tasker.taskData.isInvisible = true }
        if(enterprisePayload!==null) { tasker.taskData.isEnterprise = true; tasker.taskData.enterprisePayload = enterprisePayload }
        return await tasker.execute(this.rqdelay)
    }

    async hcaptchaenterprise(websiteURL, websiteKey, proxyInfo, userAgent=null, isInvisible=null, enterprisePayload=null){
        let tasker = new Tasker('HCaptchaEnterpriseTask', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        tasker.taskData.isEnterprise = true
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(isInvisible!==null) { tasker.taskData.isInvisible = true }
        if(enterprisePayload!==null) { tasker.taskData.isEnterprise = true; tasker.taskData.enterprisePayload = enterprisePayload }
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    async hcaptchaturbo(websiteURL, websiteKey, proxyInfo, userAgent=null, isInvisible=null, enterprisePayload=null){
        let tasker = new Tasker('HCaptchaTurboTask', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        tasker.taskData.isEnterprise = true
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(isInvisible!==null) { tasker.taskData.isInvisible = true }
        if(enterprisePayload!==null) { tasker.taskData.isEnterprise = true; tasker.taskData.enterprisePayload = enterprisePayload }
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    async hcaptchaenterpriseproxyless(websiteURL, websiteKey, userAgent=null, isInvisible=null, enterprisePayload=null){
        let tasker = new Tasker('HCaptchaEnterpriseTaskProxyLess', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        tasker.taskData.isEnterprise = true
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(isInvisible!==null) { tasker.taskData.isInvisible = true }
        if(enterprisePayload!==null) { tasker.taskData.isEnterprise = true; tasker.taskData.enterprisePayload = enterprisePayload }
        return await tasker.execute(this.rqdelay)
    }

    async hcaptchaclassification(question, base64){
        let tasker = new Tasker('HCaptchaClassification', this.apikey, this.verbose, false)
        tasker.taskData.queries = base64
        tasker.taskData.question = question
        return await tasker.execute(this.rqdelay)
    }

    /** recap **/
    async recaptchav2(websiteURL, websiteKey, proxyInfo, userAgent=null, isInvisible=null, recaptchaDataSValue=null, cookies=null){
        let tasker = new Tasker('RecaptchaV2Task', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(isInvisible!==null) { tasker.taskData.isInvisible = true }
        if(recaptchaDataSValue!==null) { tasker.taskData.recaptchaDataSValue = recaptchaDataSValue }
        if(cookies!==null) { tasker.taskData.cookies = cookies }
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    async recaptchav2proxyless(websiteURL, websiteKey, userAgent=null, isInvisible=null, recaptchaDataSValue=null, cookies=null){
        let tasker = new Tasker('RecaptchaV2TaskProxyless', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(isInvisible!==null) { tasker.taskData.isInvisible = true }
        if(recaptchaDataSValue!==null) { tasker.taskData.recaptchaDataSValue = recaptchaDataSValue }
        if(cookies!==null) { tasker.taskData.cookies = cookies }
        return await tasker.execute(this.rqdelay)
    }

    async recaptchav2enterprise(websiteURL, websiteKey, proxyInfo, userAgent=null, enterprisePayload=null, apiDomain=null, cookies=null){
        let tasker = new Tasker('RecaptchaV2EnterpriseTask', this.apikey, this.verbose);
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        if(enterprisePayload!==null) { tasker.taskData.isEnterprise = true; tasker.taskData.enterprisePayload = enterprisePayload }
        if(apiDomain!==null) { tasker.taskData.apiDomain = apiDomain }
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(cookies!==null) { tasker.taskData.cookies = cookies }
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    async recaptchav2enterpriseproxyless(websiteURL, websiteKey, userAgent=null, enterprisePayload=null, apiDomain=null, cookies=null){
        let tasker = new Tasker('RecaptchaV2EnterpriseTaskProxyless', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        if(enterprisePayload!==null) { tasker.taskData.isEnterprise = true; tasker.taskData.enterprisePayload = enterprisePayload }
        if(apiDomain!==null) { tasker.taskData.apiDomain = apiDomain }
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(cookies!==null) { tasker.taskData.cookies = cookies }
        return await tasker.execute(this.rqdelay)
    }

    async recaptchav3(websiteURL, websiteKey, proxyInfo, pageAction, minScore=null){
        let tasker = new Tasker('RecaptchaV3Task', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        tasker.taskData.pageAction = pageAction
        if(minScore!==null) { tasker.taskData.minScore = minScore }
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    async recaptchav3proxyless(websiteURL, websiteKey, pageAction, minScore=null){
        let tasker = new Tasker('RecaptchaV3TaskProxyless', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        tasker.taskData.pageAction = pageAction
        if(minScore!==null) { tasker.taskData.minScore = minScore }
        return await tasker.execute(this.rqdelay)
    }

    async recaptchav3enterprise(websiteURL, websiteKey, proxyInfo, pageAction, minScore=null, enterprisePayload=null, apiDomain=null, userAgent=null, cookies=null){
        let tasker = new Tasker('RecaptchaV3Task', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        tasker.taskData.pageAction = pageAction
        if(minScore!==null) { tasker.taskData.minScore = minScore }
        if(enterprisePayload!==null) { tasker.taskData.enterprisePayload = enterprisePayload }
        if(apiDomain!==null) { tasker.taskData.apiDomain = apiDomain }
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(cookies!==null) { tasker.taskData.cookies = cookies }
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    async recaptchav3enterpriseproxyless(websiteURL, websiteKey, pageAction, minScore=null, enterprisePayload=null, apiDomain=null, userAgent=null, cookies=null){
        let tasker = new Tasker('RecaptchaV3TaskProxyless', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        tasker.taskData.pageAction = pageAction
        if(minScore!==null) { tasker.taskData.minScore = minScore }
        if(enterprisePayload!==null) { tasker.taskData.enterprisePayload = enterprisePayload }
        if(apiDomain!==null) { tasker.taskData.apiDomain = apiDomain }
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(cookies!==null) { tasker.taskData.cookies = cookies }
        return await tasker.execute(this.rqdelay)
    }

    /** datadome **/
    async datadome(websiteURL, userAgent, captchaUrl, proxyInfo){
        let tasker = new Tasker('DataDomeSliderTask', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.captchaUrl = captchaUrl
        tasker.taskData.userAgent = userAgent
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    /** funcap **/
    async funcaptcha(websiteURL, websitePublicKey, proxyInfo, funcaptchaApiJSSubdomain, userAgent=null, data=null){
        let tasker = new Tasker('FunCaptchaTask', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websitePublicKey = websitePublicKey
        tasker.taskData.funcaptchaApiJSSubdomain = funcaptchaApiJSSubdomain
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(data!==null) { tasker.taskData.data = data }
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    async funcaptchaproxyless(websiteURL, websitePublicKey, funcaptchaApiJSSubdomain, userAgent=null, data=null){
        let tasker = new Tasker('FunCaptchaTaskProxyless', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websitePublicKey = websitePublicKey
        tasker.taskData.funcaptchaApiJSSubdomain = funcaptchaApiJSSubdomain
        if(userAgent!==null) { tasker.taskData.userAgent = userAgent }
        if(data!==null) { tasker.taskData.data = data }
        return await tasker.execute(this.rqdelay)
    }

    async funcaptchaclassification(base64, question){
        let tasker = new Tasker('FunCaptchaClassification', this.apikey, this.verbose, false)
        tasker.taskData.image = base64
        tasker.taskData.question = question
        return await tasker.execute(this.rqdelay)
    }

    /** geetest **/
    async geetest(websiteURL, gt=null, challenge=null, proxyInfo, geetestApiServerSubdomain=null, captchaId=null){
        let tasker = new Tasker('GeeTestTask', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        if(challenge!==null) { tasker.taskData.challenge = challenge }
        if(gt!==null) { tasker.taskData.gt = gt }
        if(geetestApiServerSubdomain!==null) { tasker.taskData.geetestApiServerSubdomain = geetestApiServerSubdomain }
        if(captchaId!==null) { tasker.taskData.captchaId = captchaId }
        this.attachProxy(tasker, proxyInfo)
        return await tasker.execute(this.rqdelay)
    }

    async geetestproxyless(websiteURL, gt=null, challenge=null, geetestApiServerSubdomain=null, captchaId=null){
        let tasker = new Tasker('GeeTestTaskProxyless', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        if(challenge!==null) { tasker.taskData.challenge = challenge }
        if(gt!==null) { tasker.taskData.gt = gt }
        if(geetestApiServerSubdomain!==null) { tasker.taskData.geetestApiServerSubdomain = geetestApiServerSubdomain }
        if(captchaId!==null) { tasker.taskData.captchaId = captchaId }
        return await tasker.execute(this.rqdelay)
    }

    /** anticloudflare turnstile captcha **/
    async antiturnstile(websiteURL, websiteKey, proxyInfo, metadata=null){
        let tasker = new Tasker('AntiCloudflareTask', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        tasker.taskData.websiteKey = websiteKey
        this.attachProxy(tasker, proxyInfo)
        if(metadata!==null) { tasker.taskData.metadata = metadata }
        return await tasker.execute(this.rqdelay)
    }

    /** anticloudflare challenge **/
    async anticloudflare(websiteURL, proxyInfo, metadata=null, html=null){
        let tasker = new Tasker('AntiCloudflareTask', this.apikey, this.verbose)
        tasker.taskData.websiteURL = websiteURL
        this.attachProxy(tasker, proxyInfo)
        if(metadata!==null) { tasker.taskData.metadata = metadata }
        if(html!==null) { tasker.taskData.html = html }
        return await tasker.execute(this.rqdelay)
    }

}

module.exports = CapSolver
