// capsolver-npm/src/Tasker.js
const Validation = require('./Validation')
const sleep = ms => new Promise(r => setTimeout(r, ms))
const axios = require('axios')

class Tasker {
    constructor(type, apiKey, verbose, retrieve=true) {
        this.apikey = apiKey
        this.verbose = verbose
        this.retrieve = retrieve
        this.parameters = new Validation().parameters
        this.taskData = { 'type' : type }
    }

    /**
     * executes the task
     */
    async execute(rqdelay){
        if(this.retrieve === true){
            let tasked = await this.createTask()
            if(tasked.error !== 0) return tasked
            return await this.getTaskResult(tasked['apiResponse']['taskId'], rqdelay)
        }else{
            return await this.createTask()
        }
    }

    /**
     * api.capsolver.com/createTask
     */
    async createTask(url=undefined) {
        let self = this
        this.validate(this.taskData)
        let req = { url: (url === undefined) ? 'https://api.capsolver.com/createTask' : url, headers: { }, method: 'post', data: { 'clientKey': this.apikey.toString(), 'appId': 'AF0F28E5-8245-49FD-A3FD-43D576C0E9B3', 'task': this.taskData } }
        if(this.verbose === 2 ){ console.log(req) }
        let handled = await axios(req)
            .then(async function (response) {
                if(self.verbose === 2){ console.log(response.data) }
                if(response.data['errorId'] !== 0){ return { 'error':-1, 'statusText':response.status, 'apiResponse':response.data } }
                return { 'error':0, 'statusText':response.status, 'apiResponse':response.data, 'solution':response.data.solution }
            })
            .catch(function (error) {
                if(error.response === undefined){ return error }
                if(self.verbose === 2){ console.log(error.response.data) }
                return { 'error':-1, 'statusText':error.response.status, 'apiResponse':error.response.data }
            })
        if(this.verbose === 1){ if(handled.error === 0){ console.log(`capsolver-npm: ${this.taskData.type}: [created task][${handled['apiResponse']['taskId']}]`) } else { console.log(`capsolver-npm: ${this.taskData.type}: [failed][${handled['apiResponse']}]`) } }
        if(this.verbose === 2){ console.log(handled['apiResponse']) }
        return handled
    }

    /**
     * api.capsolver.com/getTaskResult - retrieve results loop
     * @param {string} taskId - associated taskId
     * @param rqdelay
     */
    async getTaskResult(taskId, rqdelay){
        let self = this; let status = ''; let fails = 0; let handled = null
        if(taskId === undefined) return
        let requestData = { 'clientKey':this.apikey, 'taskId': taskId }
        let req = { method: 'post', url: 'https://api.capsolver.com/getTaskResult', headers: { }, data: requestData }
        while(status !== 'ready'){
            await sleep(rqdelay)
            if(fails > 10) break
            handled = await axios(req)
                .then(async function (response) {
                    if(response.data['errorId'] !== 0){
                        status = JSON.stringify(response.data)
                        return { 'error':-1, 'statusText':response.status, 'apiResponse':response.data }
                    }
                    status = response.data.status
                    return { 'error':0, 'statusText':response.status, 'apiResponse':response.data, 'solution':response.data.solution }
                })
                .catch(function (error) {
                    if(self.verbose === 2){ console.log(error.response.data) }
                    fails++
                    return { 'error':-1, 'statusText':error.response.status, 'apiResponse':error.response.data }
                })
            if(this.verbose === 1){ console.log(`capsolver-npm: [${taskId}][status: ${status}]`) }
            if(this.verbose === 2){ console.log(handled['apiResponse']) }
            if(handled.error !== 0) break
        }
        return handled
    }

    /**
     * Validate captcha task required parameters
     * @param {object} taskData - taskData schema
     */
    validate(taskData){
        const parameters = Object.keys(this.parameters).reduce((c, k) => (c[k.toLowerCase()] = this.parameters[k], c), {})
        if(Object.keys(parameters).includes(taskData.type.toLowerCase())){ // ?is a existing captcha task
            this.parameters[taskData.type].forEach(parameter => {
                if(parameter.required){
                    if(taskData[parameter.name] === undefined
                        || String(taskData[parameter.name]).length <= 0
                        || typeof taskData[parameter.name] !== parameter.type){
                        throw TypeError(`${parameter.name} must by of type ${parameter.type} and not empty.`)
                    }
                }else{
                    if(taskData[parameter.name] !== undefined || null){
                        if(String(taskData[parameter.name]).length <= 0
                            || typeof taskData[parameter.name] !== parameter.type) {
                            throw TypeError(`${parameter.name} must by of type ${parameter.type} and not empty.`)
                        }
                    }
                }
            })
        }else{
            // allow custom task types
            if(this.verbose !== 0){ console.log(`capsolver-npm: [Using a custom TaskType: ${taskData.type}]`) }
        }
        return true
    }

}

module.exports = Tasker
